jpeg.go

```diff
@@ -5,49 +5,156 @@ import (
 	"encoding/binary"
 	"errors"
 	"io"
+	"io/ioutil"
 )
 
 var headerBytes []byte = []byte{0xff, 0xd8}
 
+// scanReader reads until the scan terminating bytes are encountered
+type scanReader struct {
+	reader *io.Reader
+}
+
+func firstMarker(p []byte) int {
+	markerIndex := -1
+	// Look for an 0xff followed by anything other than 0x00 or [0xd0-0xd7] (those are RST markers)
+	for i := 0x01; i <= 0xff; i++ {
+		if i >= 0xd0 && i <= 0xd7 {
+			continue
+		}
+		markerIndex = bytes.Index(p, []byte{0xff, byte(i)})
+		if markerIndex >= 0 {
+			return markerIndex
+		}
+	}
+	return markerIndex
+}
+
+func (sr *scanReader) Read(p []byte) (n int, err error) {
+	buf := new(bytes.Buffer)
+	teedReader := io.TeeReader(*sr.reader, buf)
+	// kick the underlying reader into giving us more than just one byte because the
+	// the naive multireader will just give us one byte from the first reader in its arguments
+	if len(p) > 1 {
+		if n, err = io.ReadAtLeast(teedReader, p, 2); err != nil {
+			if errors.Is(err, io.EOF) {
+				return
+			} else if errors.Is(err, io.ErrUnexpectedEOF) {
+				// safe to just deal with on the next pass through
+				err = nil
+			} else {
+				err = &MalformedDataError{"unexpected error reading inside scan data", err}
+				return
+			}
+		}
```

Comment:
Seems like you're tolerant of 1-byte reads (based on the `else` block), so why do this extra work here?

Reply:
If you don't give preference to multi-byte reads, it'll only ever get 1-byte reads. That's because after the first read the underlying reader is a `io.MultiReader` composed of a single byte reader followed by the "tail" reader.

Instead we have to kind of "kick" the reader here to try and convince it to default to giving us multibyte reads. It's a little annoying, but also :shrug: ??

Reply:
Ah I see.  That's a funny side effect =D